### PR TITLE
config: inotify Subscription implementation.

### PR DESCRIPTION
**Pull request by htuch**
Filesystem inotify implementation of the API Subscription interface. This allows the API to be consumed on filesystem changes to files containing the JSON canonical representation of lists of ResourceType.

### CODE REVIEW

File: source/common/filesystem/subscription_impl.h [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+namespace Envoy {
+namespace Filesystem {
+
+/**
+ * Filesystem inotify implementation of the API Subscription interface. This allows the API to be
+ * consumed on filesystem changes to files containing the JSON canonical representation of
+ * lists of ResourceType.
+ */
+template <class ResourceType> class SubscriptionImpl : public Config::Subscription<ResourceType> {
```
Comment (mattklein123) [inline source/common/filesystem/subscription_impl.h]: pre-review quick nit: Feels like this should actually go in config namespace?

File: source/common/filesystem/subscription_impl.h [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+public:
+  SubscriptionImpl(Event::Dispatcher& dispatcher, const std::string& path)
+      : path_(path), watcher_(dispatcher.createFilesystemWatcher()) {}
+
+  // Config::Subscription
+  void start(const std::vector<std::string>& resources,
+             Config::SubscriptionCallbacks<ResourceType>& callbacks) override {
+    // We report all discovered resources in the watched file.
+    UNREFERENCED_PARAMETER(resources);
+    callbacks_ = &callbacks;
```
Comment (dnoe) [inline source/common/filesystem/subscription_impl.h]: Why not keep this as a reference?

File: source/common/filesystem/subscription_impl.h [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    watcher_->addWatch(path_, Watcher::Events::MovedTo, [this](uint32_t events) {
+      UNREFERENCED_PARAMETER(events);
+      refresh();
```
Comment (dnoe) [inline source/common/filesystem/subscription_impl.h]: This should be safe, because when the watcher_ goes out of scope it cleans up resources.  The case to watch out for is any case where the callback execution somehow causes this object to be destroyed. I don't think this should happen, but similar bugs have been seen in the past. It might be worth a doc comment indicating when the DeferredDeletable dance must be used in the callbacks.

File: source/common/filesystem/subscription_impl.h (hunk 1 above)
Comment (htuch) [inline source/common/filesystem/subscription_impl.h]: I've placed the `Subscription` interface in the `Config` namespace, then each of the `SubscriptionImpl` subclasses in `Grpc`, `Http`, `Filesystem` as needed. Are you suggesting a `Config` namespace within these?

Comment (mattklein123) [inline source/common/filesystem/subscription_impl.h]: Sorry maybe I wasn't paying attention, but it feels like all config impls should be inside the config namespace so they would be inside `/source/common/config/...` It feels like this would colocate a lot of related config loading code even if the transports are the same, whereas the other namespaces are for more low level stuff. 

This is really just how I logically view things. Don't feel super strongly if you feel otherwise.

Comment (htuch) [inline source/common/filesystem/subscription_impl.h]: Let me take an AI to do this as a separate PR once this one is in, I don't mind the cleanup, but it's best done once globally.

Comment (mattklein123) [inline source/common/filesystem/subscription_impl.h]: OK sounds good I will review as is in the current location.

File: source/common/filesystem/subscription_impl.h (hunk 2 above)
Comment (htuch) [inline source/common/filesystem/subscription_impl.h]: If it's a reference, it needs to be initialized in the constructor, but we don't know the value until `start` is invoked.

File: source/common/filesystem/subscription_impl.h (hunk 3 above)
Comment (htuch) [inline source/common/filesystem/subscription_impl.h]: https://github.com/lyft/envoy/blob/master/include/envoy/config/subscription.h#L38 describes the lifetime semantics of callbacks. They need to be around for the lifetime of the Subscription object. It would seem very unusual to have the live-forever Subscription object cleaned up during a callback. I'm not against adding more comments, what would you suggest?

File: source/common/filesystem/subscription_impl.h (hunk 2 above)
Comment (dnoe) [inline source/common/filesystem/subscription_impl.h]: Got it (I suppose you could solve this with Optional too, but that can get ugly). It might be good to add a comment on the member that it's a non-owned pointer.

File: source/common/filesystem/subscription_impl.h (hunk 3 above)
Comment (dnoe) [inline source/common/filesystem/subscription_impl.h]: As long as the subscription object always lives as long as the Envoy process I think there's no issue.  Some kind of convoluted error handling is the way I see it could happen. Something simple to make this 100% clear for the comments there:

"The callback must not result in the deletion of the Subscription object"

File: test/common/filesystem/subscription_impl_test.cc [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+  }
+
+  const std::string path_;
+  Event::DispatcherImpl dispatcher_;
+  Config::MockSubscriptionCallbacks<envoy::api::v2::ClusterLoadAssignment> callbacks_;
+  EdsSubscriptionImpl subscription_;
+};
+
+// Validate basic inotify succeeds.
+TEST_F(FilesystemSubscriptionImplTest, Basic) {
```
Comment (mattklein123) [inline test/common/filesystem/subscription_impl_test.cc]: IIRC, a lot of these tests are basically the same as the other subscription impl. Is there any ability to share during the refactor to move all code into config? NBD but thought I would mention.

Comment (htuch) [inline test/common/filesystem/subscription_impl_test.cc]: Yeah. I was waiting to see how many of these tests would make sense across the implementations and for the interface to encapsulate the individual implementations to materialize at the test level. This now seems apparent, I'll take an AI to turn this into a parameterized test (TEST_P) later on.
